src/mono/wasm/runtime/method-binding.ts

```diff
@@ -241,28 +240,29 @@ export function _compile_converter_for_marshal_string(args_marshal: string/*Args
         const valueKey = "value" + i;
 
         const argKey = "arg" + i;
+        const offsetText = `(indirectStart + ${indirectLocalOffset})`;
         argumentNames.push(argKey);
 
         if (step.convert_root) {
-            body.push("if (!rootBuffer) throw new Error('no root buffer provided');");
-            // FIXME: Optimize this!!!
-            if (!converter.scratchValueRoot)
-                closure.scratchValueRoot = converter.scratchValueRoot = mono_wasm_new_root<MonoObject>();
+            mono_assert(!step.indirect, "converter step cannot both be rooted and indirect");
+            if (!converter.scratchValueRoot) {
+                // HACK: new_external_root rightly won't accept a null address
+                const dummyAddress = Module.stackSave();
+                converter.scratchValueRoot = mono_wasm_new_external_root<MonoObject>(dummyAddress);
```

Comment:
I guess we don't nee to allocate root object at all here. Simple helper function would suffice. Since we generate all code which is using it, we don't need the interface, right ?

Reply:
I think the converters this calls need a root, so we'd need to change the whole API. And some of them are public